File: 17298.py
from sys import stdin

input = stdin.readline


def solve():
    N = int(input())
    arr = [*map(int, input().split())]
    stack = []
    ans = [-1] * N
    for i, num in enumerate(arr):
        while stack and arr[stack[-1]] < num:
            ans[stack.pop()] = num
        stack.append(i)

    print(' '.join(map(str, ans)))


if __name__ == '__main__':
    solve()

File: 17298.js
const fs = require('fs');

function solve() {
	const input = fs.readFileSync('/dev/stdin').toString().split('\n');
	let inputIdx = 0;
	const N = +input[inputIdx++];
	const arr = input[inputIdx++].split(' ').map((v) => +v);
	const ans = [];
	ans.length = N;
	const stack = [];

	for (let i = 0; i < N; i++) {
		while (stack.length && arr[stack[stack.length - 1]] < arr[i]) {
			ans[stack.pop()] = arr[i];
		}
		stack.push(i);
	}
	for (const i of stack) ans[i] = -1;
	console.log(ans.join(' '));
}

solve();
